<p style="text-align:center">
    <a href="https://skills.network/?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkPY0220ENSkillsNetwork900-2022-01-01" target="_blank">
    <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/assets/logos/SN_web_lightmode.png" width="200" alt="Skills Network Logo">
    </a>
</p>


<h1>Extracting and Visualizing Stock Data</h1>
<h2>Description</h2>


Extracting essential data from a dataset and displaying it is a necessary part of data science; therefore individuals can make correct decisions based on the data. In this assignment, you will extract some stock data, you will then display this data in a graph.


<h2>Table of Contents</h2>
<div class="alert alert-block alert-info" style="margin-top: 20px">
    <ul>
        <li>Define a Function that Makes a Graph</li>
        <li>Question 1: Use yfinance to Extract Stock Data</li>
        <li>Question 2: Use Webscraping to Extract Tesla Revenue Data</li>
        <li>Question 3: Use yfinance to Extract Stock Data</li>
        <li>Question 4: Use Webscraping to Extract GME Revenue Data</li>
        <li>Question 5: Plot Tesla Stock Graph</li>
        <li>Question 6: Plot GameStop Stock Graph</li>
    </ul>
<p>
    Estimated Time Needed: <strong>30 min</strong></p>
</div>

<hr>


***Note***:- If you are working in IBM Cloud Watson Studio, please replace the command for installing nbformat from `!pip install nbformat==4.2.0` to simply `!pip install nbformat`


In [1]:
!pip install yfinance==0.1.67
!mamba install bs4==4.10.0 -y
!pip install nbformat==4.2.0



                  __    __    __    __
                 /  \  /  \  /  \  /  \
                /    \/    \/    \/    \
███████████████/  /██/  /██/  /██/  /████████████████████████
              /  / \   / \   / \   / \  \____
             /  /   \_/   \_/   \_/   \    o \__,
            / _/                       \_____/  `
            |/
        ███╗   ███╗ █████╗ ███╗   ███╗██████╗  █████╗
        ████╗ ████║██╔══██╗████╗ ████║██╔══██╗██╔══██╗
        ██╔████╔██║███████║██╔████╔██║██████╔╝███████║
        ██║╚██╔╝██║██╔══██║██║╚██╔╝██║██╔══██╗██╔══██║
        ██║ ╚═╝ ██║██║  ██║██║ ╚═╝ ██║██████╔╝██║  ██║
        ╚═╝     ╚═╝╚═╝  ╚═╝╚═╝     ╚═╝╚═════╝ ╚═╝  ╚═╝

        mamba (0.27.0) supported by @QuantStack

        GitHub:  https://github.com/mamba-org/mamba
        Twitter: https://twitter.com/QuantStack

█████████████████████████████████████████████████████████████


Looking for: ['bs4==4.10.0']

[+] 0.0s
opt/ibm/custom-channels/meta-wscloud/linux-64      129.0 B @   1.6MB/s  0

Extracting   (3) ━━━━╸━━━━━━━━━━━━━━━╸━━       0 ibm-wsrt-py310t2-def       0.1sbeautifulsoup4                                      86.6kB @ 379.8kB/s  0.2s
[+] 0.3s
Extracting   (4) ━━━━━╸━━━━━━━━━━━━━━━╸━       0 ibm-wsrt-py310t2-def       0.2s[+] 0.4s
Extracting   (4) ━━━━━━╸━━━━━━━━━━━━━━━━       0 ibm-wsrt-py310t2-def       0.3s[+] 0.5s
Extracting   (4) ━━━━━━━╸━━━━━━━━━━━━━━━       0 beautifulsoup4             0.4s[+] 0.6s
Extracting   (4) ━━━━━━━━╸━━━━━━━━━━━━━━       0 beautifulsoup4             0.5s[+] 0.7s
Extracting   (4) ━━━━━━━━━╸━━━━━━━━━━━━━       0 beautifulsoup4             0.6s[+] 0.8s
Extracting   (4) ━━━━━━━━━━╸━━━━━━━━━━━━       0 beautifulsoup4             0.7s[+] 0.9s
Extracting   (4) ━━━━━━━━━━━╸━━━━━━━━━━━       0 bs4                        0.8s[+] 1.0s
Extracting   (4) ━━━━━━━━━━━━╸━━━━━━━━━━       0 bs4                        0.9s[+] 1.1s
Extracting   (4) ━━━━━━━╸━━━━━━━━━━━━━━━       0 bs4                        1.0s[+] 1.2s
Extracting   (4) ━━━━━━━━━╸━━━━━━

Extracting   (4) ━╸━━━━━━━━━━━━━━━╸━━━━━       0 beautifulsoup4             3.9s[+] 4.1s
Extracting   (4) ━━╸━━━━━━━━━━━━━━━╸━━━━       0 bs4                        4.0s[+] 4.2s
Extracting   (4) ━━━━╸━━━━━━━━━━━━━━━╸━━       0 bs4                        4.1s[+] 4.3s
Extracting   (4) ━━━━━╸━━━━━━━━━━━━━━━╸━       0 bs4                        4.2s[+] 4.4s
Extracting   (4) ━━━━━━╸━━━━━━━━━━━━━━━━       0 bs4                        4.3s[+] 4.5s
Extracting   (4) ━━━━━━━╸━━━━━━━━━━━━━━━       0 ca-certificates            4.4s[+] 4.6s
Extracting   (4) ━━━━━━━━╸━━━━━━━━━━━━━━       0 ca-certificates            4.5s[+] 4.7s
Extracting   (4) ━━━━━━━━━╸━━━━━━━━━━━━━       0 ca-certificates            4.6s[+] 4.8s
Extracting   (4) ━━━━━━━━━━╸━━━━━━━━━━━━       0 ca-certificates            4.7s[+] 4.9s
Extracting   (4) ━━━━━━━━━━━╸━━━━━━━━━━━       0 ibm-wsrt-py310t2-def       4.8s[+] 5.0s
Extracting   (4) ━━━━━━━━━━━━╸━━━━━━━━━━       0 ibm-wsrt-py310t2-def       4.9s[+] 5.1s
Extracting   (4) ━━━━

Extracting   (4) ━━━━━━━━━━━━━━━╸━━━━━━━       0 ca-certificates            7.7s[+] 7.9s
Extracting   (4) ╸━━━━━━━━━━━━━━━╸━━━━━━       0 ca-certificates            7.8s[+] 8.0s
Extracting   (4) ━╸━━━━━━━━━━━━━━━╸━━━━━       0 ca-certificates            7.9s[+] 8.1s
Extracting   (4) ━━╸━━━━━━━━━━━━━━━╸━━━━       0 ibm-wsrt-py310t2-def       8.0s[+] 8.2s
Extracting   (4) ━━━━╸━━━━━━━━━━━━━━━╸━━       0 ibm-wsrt-py310t2-def       8.1s[+] 8.3s
Extracting   (4) ━━━━━╸━━━━━━━━━━━━━━━╸━       0 ibm-wsrt-py310t2-def       8.2s[+] 8.4s
Extracting   (4) ━━━━━━╸━━━━━━━━━━━━━━━━       0 ibm-wsrt-py310t2-def       8.3s[+] 8.5s
Extracting   (4) ━━━━━━━╸━━━━━━━━━━━━━━━       0 beautifulsoup4             8.4s[+] 8.6s
Extracting   (4) ━━━━━━━━╸━━━━━━━━━━━━━━       0 beautifulsoup4             8.5s[+] 8.7s
Extracting   (4) ━━━━━━━━━╸━━━━━━━━━━━━━       0 beautifulsoup4             8.6s[+] 8.8s
Extracting   (4) ━━━━━━━━━━╸━━━━━━━━━━━━       0 beautifulsoup4             8.7s[+] 8.9s
Extracting   (4) ━━━━

Extracting   (4) ━━━━━━━━━━━━━╸━━━━━━━━━       0 ibm-wsrt-py310t2-def      11.5s[+] 11.7s
Extracting   (4) ━━━━━━━━━━━━━━╸━━━━━━━━       0 beautifulsoup4            11.6s[+] 11.8s
Extracting   (4) ━━━━━━━━━━━━━━━╸━━━━━━━       0 beautifulsoup4            11.7s[+] 11.9s
Extracting   (4) ╸━━━━━━━━━━━━━━━╸━━━━━━       0 beautifulsoup4            11.8s[+] 12.0s
Extracting   (4) ━╸━━━━━━━━━━━━━━━╸━━━━━       0 beautifulsoup4            11.9s[+] 12.1s
Extracting   (4) ━━╸━━━━━━━━━━━━━━━╸━━━━       0 bs4                       12.0s[+] 12.2s
Extracting   (4) ━━━━╸━━━━━━━━━━━━━━━╸━━       0 bs4                       12.1s[+] 12.3s
Extracting   (4) ━━━━━╸━━━━━━━━━━━━━━━╸━       0 bs4                       12.2s[+] 12.4s
Extracting   (4) ━━━━━━╸━━━━━━━━━━━━━━━━       0 bs4                       12.3s[+] 12.5s
Extracting   (4) ━━━━━━━╸━━━━━━━━━━━━━━━       0 ca-certificates           12.4s[+] 12.6s
Extracting   (4) ━━━━━━━━╸━━━━━━━━━━━━━━       0 ca-certificates           12.5s[+] 12.7s
Extracting

Extracting   (4) ━━━━━━━━━━━╸━━━━━━━━━━━       0 bs4                       15.3s[+] 15.5s
Extracting   (4) ━━━━━━━━━━━━╸━━━━━━━━━━       0 bs4                       15.4s[+] 15.6s
Extracting   (4) ━━━━━━━━━━━━━╸━━━━━━━━━       0 bs4                       15.5s[+] 15.7s
Extracting   (4) ━━━━━━━━━━━━━━╸━━━━━━━━       0 ca-certificates           15.6s[+] 15.8s
Extracting   (4) ━━━━━━━━━━━━━━━╸━━━━━━━       0 ca-certificates           15.7s[+] 15.9s
Extracting   (4) ╸━━━━━━━━━━━━━━━╸━━━━━━       0 ca-certificates           15.8s[+] 16.0s
Extracting   (4) ━╸━━━━━━━━━━━━━━━╸━━━━━       0 ca-certificates           15.9s[+] 16.1s
Extracting   (4) ━━╸━━━━━━━━━━━━━━━╸━━━━       0 ibm-wsrt-py310t2-def      16.0s[+] 16.2s
Extracting   (4) ━━━━╸━━━━━━━━━━━━━━━╸━━       0 ibm-wsrt-py310t2-def      16.1s[+] 16.3s
Extracting   (4) ━━━━━╸━━━━━━━━━━━━━━━╸━       0 ibm-wsrt-py310t2-def      16.2s[+] 16.4s
Extracting   (4) ━━━━━━╸━━━━━━━━━━━━━━━━       0 ibm-wsrt-py310t2-def      16.3s[+] 16.5s
Extracting

Extracting   (4) ━━━━━━━━━╸━━━━━━━━━━━━━       0 ca-certificates           19.1s[+] 19.3s
Extracting   (4) ━━━━━━━━━━╸━━━━━━━━━━━━       0 ibm-wsrt-py310t2-def      19.2s[+] 19.4s
Extracting   (4) ━━━━━━━━━━━╸━━━━━━━━━━━       0 ibm-wsrt-py310t2-def      19.3s[+] 19.5s
Extracting   (4) ━━━━━━━━━━━━╸━━━━━━━━━━       0 ibm-wsrt-py310t2-def      19.4s[+] 19.6s
Extracting   (4) ━━━━━━━━━━━━━╸━━━━━━━━━       0 ibm-wsrt-py310t2-def      19.5s[+] 19.7s
Extracting   (4) ━━━━━━━━━━━━━━╸━━━━━━━━       0 beautifulsoup4            19.6s[+] 19.8s
Extracting   (4) ━━━━━━━━━━━━━━━╸━━━━━━━       0 beautifulsoup4            19.7s[+] 19.9s
Extracting   (4) ╸━━━━━━━━━━━━━━━╸━━━━━━       0 beautifulsoup4            19.8s[+] 20.0s
Extracting   (4) ━╸━━━━━━━━━━━━━━━╸━━━━━       0 beautifulsoup4            19.9s[+] 20.1s
Extracting   (4) ━━╸━━━━━━━━━━━━━━━╸━━━━       0 bs4                       20.0s[+] 20.2s
Extracting   (4) ━━━━╸━━━━━━━━━━━━━━━╸━━       0 bs4                       20.1s[+] 20.3s
Extracting

  Attempting uninstall: nbformat
    Found existing installation: nbformat 5.3.0
    Uninstalling nbformat-5.3.0:
      Successfully uninstalled nbformat-5.3.0


In [3]:
import yfinance as yf
import pandas as pd
import requests
from bs4 import BeautifulSoup
import plotly.graph_objects as go
from plotly.subplots import make_subplots

## Define Graphing Function


In this section, we define the function `make_graph`. You don't have to know how the function works, you should only care about the inputs. It takes a dataframe with stock data (dataframe must contain Date and Close columns), a dataframe with revenue data (dataframe must contain Date and Revenue columns), and the name of the stock.


In [4]:
def make_graph(stock_data, revenue_data, stock):
    fig = make_subplots(rows=2, cols=1, shared_xaxes=True, subplot_titles=("Historical Share Price", "Historical Revenue"), vertical_spacing = .3)
    stock_data_specific = stock_data[stock_data.Date <= '2021--06-14']
    revenue_data_specific = revenue_data[revenue_data.Date <= '2021-04-30']
    fig.add_trace(go.Scatter(x=pd.to_datetime(stock_data_specific.Date, infer_datetime_format=True), y=stock_data_specific.Close.astype("float"), name="Share Price"), row=1, col=1)
    fig.add_trace(go.Scatter(x=pd.to_datetime(revenue_data_specific.Date, infer_datetime_format=True), y=revenue_data_specific.Revenue.astype("float"), name="Revenue"), row=2, col=1)
    fig.update_xaxes(title_text="Date", row=1, col=1)
    fig.update_xaxes(title_text="Date", row=2, col=1)
    fig.update_yaxes(title_text="Price ($US)", row=1, col=1)
    fig.update_yaxes(title_text="Revenue ($US Millions)", row=2, col=1)
    fig.update_layout(showlegend=False,
    height=900,
    title=stock,
    xaxis_rangeslider_visible=True)
    fig.show()

## Question 1: Use yfinance to Extract Stock Data


Using the `Ticker` function enter the ticker symbol of the stock we want to extract data on to create a ticker object. The stock is Tesla and its ticker symbol is `TSLA`.


In [26]:
Tesla = yf.Ticker("TSLA")

Using the ticker object and the function `history` extract stock information and save it in a dataframe named `tesla_data`. Set the `period` parameter to `max` so we get information for the maximum amount of time.


In [27]:
tesla_data = Tesla.history(period="max")

**Reset the index** using the `reset_index(inplace=True)` function on the tesla_data DataFrame and display the first five rows of the `tesla_data` dataframe using the `head` function. Take a screenshot of the results and code from the beginning of Question 1 to the results below.


In [24]:
tesla_data.reset_index(inplace=True)
tesla_data.head()

Date      Open      High       Low     Close     Volume  Dividends  \
0 2010-06-29  1.266667  1.666667  1.169333  1.592667  281494500          0   
1 2010-06-30  1.719333  2.028000  1.553333  1.588667  257806500          0   
2 2010-07-01  1.666667  1.728000  1.351333  1.464000  123282000          0   
3 2010-07-02  1.533333  1.540000  1.247333  1.280000   77097000          0   
4 2010-07-06  1.333333  1.333333  1.055333  1.074000  103003500          0   

   Stock Splits  
0           0.0  
1           0.0  
2           0.0  
3           0.0  
4           0.0

## Question 2: Use Webscraping to Extract Tesla Revenue Data


Use the `requests` library to download the webpage https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-PY0220EN-SkillsNetwork/labs/project/revenue.htm Save the text of the response as a variable named `html_data`.


In [8]:
url = " https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-PY0220EN-SkillsNetwork/labs/project/revenue.htm"
html_data  = requests.get(url).text

Parse the html data using `beautiful_soup`.


In [9]:
soup = BeautifulSoup(html_data,'lxml')

Using `BeautifulSoup` or the `read_html` function extract the table with `Tesla Revenue` and store it into a dataframe named `tesla_revenue`. The dataframe should have columns `Date` and `Revenue`.


<details><summary>Click here if you need help locating the table</summary>

```
    
Below is the code to isolate the table, you will now need to loop through the rows and columns like in the previous lab
    
soup.find_all("tbody")[1]
    
If you want to use the read_html function the table is located at index 1


```

</details>


In [10]:
read_html_pandas_data = pd.read_html(str(soup))

In [11]:
tesla_revenue = read_html_pandas_data[0]

tesla_revenue.head()

Tesla Annual Revenue(Millions of US $)  \
0                                    2021   
1                                    2020   
2                                    2019   
3                                    2018   
4                                    2017   

  Tesla Annual Revenue(Millions of US $).1  
0                                  $53,823  
1                                  $31,536  
2                                  $24,578  
3                                  $21,461  
4                                  $11,759

In [28]:
tesla_revenue.rename(columns={'Tesla Annual Revenue(Millions of US $)': 'Date'}, inplace=True)

In [29]:
tesla_revenue.rename(columns={'Tesla Annual Revenue(Millions of US $).1': 'Revenue'}, inplace=True)

In [35]:
tesla_revenue.head()

Date Revenue
0  2021   53823
1  2020   31536
2  2019   24578
3  2018   21461
4  2017   11759

Execute the following line to remove the comma and dollar sign from the `Revenue` column. 


In [13]:
tesla_revenue["Tesla Annual Revenue(Millions of US $).1"] = tesla_revenue['Tesla Annual Revenue(Millions of US $).1'].str.replace(',|\$',"")

/tmp/wsuser/ipykernel_166/3812067121.py:1: FutureWarning: The default value of regex will change from True to False in a future version.
  tesla_revenue["Tesla Annual Revenue(Millions of US $).1"] = tesla_revenue['Tesla Annual Revenue(Millions of US $).1'].str.replace(',|\$',"")


Execute the following lines to remove an null or empty strings in the Revenue column.


In [31]:
tesla_revenue.dropna(inplace=True)

tesla_revenue = tesla_revenue[tesla_revenue['Revenue'] != ""]

Display the last 5 row of the `tesla_revenue` dataframe using the `tail` function. Take a screenshot of the results.


In [32]:
tesla_revenue.tail()

Date Revenue
8   2013    2013
9   2012     413
10  2011     204
11  2010     117
12  2009     112

## Question 3: Use yfinance to Extract Stock Data


Using the `Ticker` function enter the ticker symbol of the stock we want to extract data on to create a ticker object. The stock is GameStop and its ticker symbol is `GME`.


In [15]:
 GameStop = yf.Ticker("GME")

Using the ticker object and the function `history` extract stock information and save it in a dataframe named `gme_data`. Set the `period` parameter to `max` so we get information for the maximum amount of time.


In [16]:
gme_data =  GameStop.history(period="max")

**Reset the index** using the `reset_index(inplace=True)` function on the gme_data DataFrame and display the first five rows of the `gme_data` dataframe using the `head` function. Take a screenshot of the results and code from the beginning of Question 3 to the results below.


In [17]:
gme_data.reset_index(inplace=True)
gme_data.head()

Date      Open      High       Low     Close    Volume  Dividends  \
0 2002-02-13  1.620129  1.693350  1.603296  1.691667  76216000        0.0   
1 2002-02-14  1.712707  1.716074  1.670626  1.683250  11021600        0.0   
2 2002-02-15  1.683250  1.687458  1.658002  1.674834   8389600        0.0   
3 2002-02-19  1.666418  1.666418  1.578047  1.607504   7410400        0.0   
4 2002-02-20  1.615920  1.662209  1.603295  1.662209   6892800        0.0   

   Stock Splits  
0           0.0  
1           0.0  
2           0.0  
3           0.0  
4           0.0

## Question 4: Use Webscraping to Extract GME Revenue Data


Use the `requests` library to download the webpage https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-PY0220EN-SkillsNetwork/labs/project/stock.html. Save the text of the response as a variable named `html_data`.


In [42]:
url1 ="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-PY0220EN-SkillsNetwork/labs/project/stock.html"
html1_data  = requests.get(url1).text

Parse the html data using `beautiful_soup`.


In [43]:
soup = BeautifulSoup(html1_data,'lxml')

Using `BeautifulSoup` or the `read_html` function extract the table with `GameStop Revenue` and store it into a dataframe named `gme_revenue`. The dataframe should have columns `Date` and `Revenue`. Make sure the comma and dollar sign is removed from the `Revenue` column using a method similar to what you did in Question 2.


<details><summary>Click here if you need help locating the table</summary>

```
    
Below is the code to isolate the table, you will now need to loop through the rows and columns like in the previous lab
    
soup.find_all("tbody")[1]
    
If you want to use the read_html function the table is located at index 1


```

</details>


In [44]:
read_html_pandas_data = pd.read_html(str(soup))
gme_revenue = read_html_pandas_data[0]

gme_revenue.head()

GameStop Annual Revenue(Millions of US $)  \
0                                       2020   
1                                       2019   
2                                       2018   
3                                       2017   
4                                       2016   

  GameStop Annual Revenue(Millions of US $).1  
0                                      $6,466  
1                                      $8,285  
2                                      $8,547  
3                                      $7,965  
4                                      $9,364

In [54]:
tesla_revenue["GameStop Annual Revenue(Millions of US $).1"] = tesla_revenue['GameStop Annual Revenue(Millions of US $).1'].str.replace(',|\$',"")

KeyError: 'GameStop Annual Revenue(Millions of US $).1'

In [55]:
gme_revenue.rename(columns={'GameStop Annual Revenue(Millions of US $)': 'Date'}, inplace=True)

In [56]:
gme_revenue.rename(columns={'GameStop Annual Revenue(Millions of US $).1': 'Revenue'}, inplace=True)

In [57]:
tesla_revenue["Revenue"] = tesla_revenue['Revenue'].str.replace(',|\$',"")

/tmp/wsuser/ipykernel_166/349343550.py:1: FutureWarning:

The default value of regex will change from True to False in a future version.



Display the last five rows of the `gme_revenue` dataframe using the `tail` function. Take a screenshot of the results.


In [49]:
gme_revenue.tail()

Date Revenue
11  2009  $8,806
12  2008  $7,094
13  2007  $5,319
14  2006  $3,092
15  2005  $1,843

## Question 5: Plot Tesla Stock Graph


Use the `make_graph` function to graph the Tesla Stock Data, also provide a title for the graph. The structure to call the `make_graph` function is `make_graph(tesla_data, tesla_revenue, 'Tesla')`. Note the graph will only show data upto June 2021.


In [51]:
make_graph(tesla_data, tesla_revenue, 'Tesla')

AttributeError: 'DataFrame' object has no attribute 'Date'

## Question 6: Plot GameStop Stock Graph


Use the `make_graph` function to graph the GameStop Stock Data, also provide a title for the graph. The structure to call the `make_graph` function is `make_graph(gme_data, gme_revenue, 'GameStop')`. Note the graph will only show data upto June 2021.


In [53]:
make_graph(gme_data, gme_revenue, 'GameStop')

TypeError: Invalid comparison between dtype=int64 and str

<h2>About the Authors:</h2> 

<a href="https://www.linkedin.com/in/joseph-s-50398b136/">Joseph Santarcangelo</a> has a PhD in Electrical Engineering, his research focused on using machine learning, signal processing, and computer vision to determine how videos impact human cognition. Joseph has been working for IBM since he completed his PhD.

Azim Hirjani


## Change Log

| Date (YYYY-MM-DD) | Version | Changed By    | Change Description        |
| ----------------- | ------- | ------------- | ------------------------- |
| 2022-02-28        | 1.2     | Lakshmi Holla | Changed the URL of GameStop |
| 2020-11-10        | 1.1     | Malika Singla | Deleted the Optional part |
| 2020-08-27        | 1.0     | Malika Singla | Added lab to GitLab       |

<hr>

## <h3 align="center"> © IBM Corporation 2020. All rights reserved. <h3/>

<p>
